src/main/java/com/microsoft/azure/relay/HybridConnectionListener.java

```diff
@@ -676,34 +679,42 @@ public void setOnlineHandler(BiConsumer<Object, Object[]> onOnline) {
 							this.address.getPath(), this.address.getQuery(), HybridConnectionConstants.Actions.LISTEN,
 							trackingId);
 
-				return delayTask.thenCompose((res) -> {
-					synchronized (this.thisLock) {
-						return this.webSocket.connectAsync(websocketUri, timeout, config);
-					}
-				}).thenRun(() -> this.onOnline());
+				ClientWebSocket webSocket = new ClientWebSocket(this.listener.trackingContext, EXECUTOR);
+				return delayTask.thenCompose(($void) -> {
+					return webSocket.connectAsync(websocketUri, timeout, config).thenApply(($void2) -> webSocket)
+						.whenComplete((ws, err) -> {
+							if (err == null) {
+								this.onOnline();
+							}
+						});
```

Comment:
I don't think this .whenComplete block is necessary and the "thenApply" should be enough. If there is an exception from websocket connecting, the "thenApply" wouldn't be run anyways.

Reply:
That sounds good.

---
In reply to: [258635262](https://github.com/Azure/azure-relay-java/pull/17#discussion_r258635262) [](ancestors = 258635262)

Reply:
Fixed in iteration 3 (iteration 2 was a merge with master).

---
In reply to: [258635798](https://github.com/Azure/azure-relay-java/pull/17#discussion_r258635798) [](ancestors = 258635798,258635262)